---
title:  "[E-commerce App] CircuitBreaker와 Resilience4J 사용 "

categories:
  - SpringCloud
tags:
  - [SpringCloud]

toc: true
toc_sticky: true
 
date: 2022-10-12 01:13:20
last_modified_at: 2022-10-12 01:13:23
---
마이크로 서비스 간의 통신 시 발생할 수 있는 대표적인 오류와 그것을 해결할 수 있는 방법에 대해 알아보자.

# Microservice 통신 시 연쇄 오류
아래 사진을 보면, `user-service`를 호출했을 때, 500 Error가 발생했는데, <br>
`trace`를 보게 되면 그 원인은 `order-service`에 있다는 것을 알 수 있다.<br>
![스크린샷 2022-10-12 오전 2 09 28](https://user-images.githubusercontent.com/59405576/195156131-e90ceaf1-10f3-4ded-b1b3-0f83432922e5.png){: width="700" height="700"}<br><br>
`user-service`에서 `order-service`를 호출하고, `order-service`에서 `catalog-service`를 호출한다고 가정하자.<br>
만약 `order-service`나 `catalog-service`에서 오류가 발생했을 때, <br>
해당하는 마이크로 서비스로 더이상 요청을 전달하지 않아야 한다.<br><br>
따라서 Feign Client 측에서는 에러가 발생했을 때, 에러를 대신할 수 있는 default 값이라던가 정상적인 데이터처럼 보여질 수 있는 다른 데이터 값을 응답할 준비가 되어있어야 한다.<br>
즉, 이는 `user-service`의 문제가 아니므로 `user-service`에서는 500 Error가 아닌 200 OK를 반환해야 한다는 것이다.<br>
![스크린샷 2022-10-12 오전 2 10 42](https://user-images.githubusercontent.com/59405576/195156360-06b92f1d-1949-406a-9caa-fd5f3332002e.png){: width="700" height="700"}<br><br>
CuircuitBreaker란, 문제가 생긴 서비스를 더이상 사용하지 않도록 막아주고, <br>
문제가 생긴 서비스가 정상적으로 복구 되어 사용할 수 있는 상태가 되었을 때 다시 사용할 수 있도록 정상적인 흐름으로 바꿔주는 장치이다.

# CuircuitBreaker
- 장애가 발생하는 서비스에 반복적인 호출이 되지 못하게 차단
- 특정 서비스가 정상적으로 동작하지 않을 경우다른 기능으로 대체 수행 -> 장애 회피

![스크린샷 2022-10-12 오전 2 19 48](https://user-images.githubusercontent.com/59405576/195157893-1ce936bc-f021-487d-88f8-44966854f1f0.png){: width="300" height="300"} ![스크린샷 2022-10-12 오전 2 22 07](https://user-images.githubusercontent.com/59405576/195158320-fdf0a2f0-9934-4552-a89f-5ed5c49d7d79.png){: width="500" height="500"}<br>
정상적으로 다른 마이크로 서비스를 사용할 수 있을 때, Circuit Breaker는 closed 되어있고,<br>
어떤 마이크로 서비스에 문제가 발생했을 때, Circuit Breaker가 open 상태가 되고, <br>
그러면 클라이언트의 요청을 더이상 마이크로 서비스에게 전달하지 않고, Circuit Breaker에서 자체적으로 기본 값이라던가 우회할 수 있는 값을 가지고 리턴 시켜준다.<br><br>
지금까지 만들었던 마이크로 서비스에 이러한 Circuit Breaker를 추가함으로써, <br>
하나의 마이크로 서비스에서 발생한 문제가 연쇄적으로 다른 마이크로 서비스에 전달되는 것을 막을 수 있다. <br>
-> 문제가 발생한 마이크로 서비스를 제외한 정상적인 마이크로 서비스는, 정상적으로 응답할 수 있다.

## Resilience4j
2019년도 이전까지는 CircuitBreaker를 사용하기 위해서 Spring Cloud Netflix Hystrix 라이브러리를 사용할 수 있었지만,<br>
Spring boot 2.4 버전에서는 Hystrix 라이브러리가 더이상 제공되지 않기 때문에 대체 할 수 있는 라이브러리로 바꿔야 한다.<br>
그것이 바로 Resilience4j이다.

# 문제 확인
문제를 확인해보기 위해 아래 순서대로 서버를 기동하자
- eureka 서버
- rabbitmq 서버
- configuration 서버 (`config-service`)
- gateway 서버 (`gateway-service`)
- `user-service`

포스트맨에서 회원가입 - 로그인 순서대로 진행하자.<br>
로그인 후 발금받은 토큰과 `userId`를 이용해 사용자의 상세정보를 요청하면, 다음과 같은 에러가 발생한다.<br>
![스크린샷 2022-10-13 오전 9 48 56](https://user-images.githubusercontent.com/59405576/195473756-884dc65c-e900-46d9-9552-a021766e9798.png)<br>
이는 사용자 단건 조회 시, 사용자의 주문 목록도 가져오는데,<br>
현재 `order-service`가 기동중이지 않아 발생한 오류이다.<br><br>
이제 CuircuitBreaker를 구현해보자.

# `user-service`
## `pom.xml`
```xml
<!-- Resilience4j -->
<dependency>
    <groupId>org.springframework.cloud</groupId>
    <artifactId>spring-cloud-starter-circuitbreaker-resilience4j</artifactId>
</dependency>
```

## `UserServiceImpl.java`
```java
@Service
@Slf4j
public class UserServiceImpl implements UserService {
    ...
    @Override
    public UserDto getUserByUserId(String userId) {

        UserEntity userEntity = userRepository.findByUserId(userId);

        if (userEntity == null) {
            throw new UsernameNotFoundException("User not found");
        }

        UserDto userDto = new ModelMapper().map(userEntity, UserDto.class);

        // 주문 생성
        /* ErrorDecoder */
        // List<ResponseOrder> orderList = orderServiceClient.getOrders(userId);

        /* 🌟 CircuitBreaker */
        CircuitBreaker circuitbreaker = circuitBreakerFactory.create("circuitbreaker");
        List<ResponseOrder> orderList = circuitbreaker.run(() -> orderServiceClient.getOrders(userId),
            throwable -> new ArrayList<>());

        userDto.setOrders(orderList);

        return userDto;
    }
}
```

## 테스트
이전과 마찬가지로 `order-service`를 기동하지 않은 채로 "사용자 단건 조회"를 요청해보자.<br>
![스크린샷 2022-10-13 오전 10 06 57](https://user-images.githubusercontent.com/59405576/195475585-93dc2611-f8ea-4acc-8de5-58d2b7bc22a8.png)<br>
이번에는 500 에러가 발생하지 않고 사용자의 기본 정보는 가져오되, 주문 내역만 가지고 오지 못하는 것을 확인할 수 있다.<br><br>
참고로 인텔리제이의 콘솔에서는 아래와 같이 에러 로그가 찍힌다.<br>
![스크린샷 2022-10-13 오전 10 09 16](https://user-images.githubusercontent.com/59405576/195475779-c6872127-273d-42e2-886e-388b4772bfd1.png)<br><br>
만약, `order-service`를 기동한 뒤에 다시 테스트해보면, 그때는 사용자의 주문 내역까지도 정상적으로 가져올 것이다.<br>
이번에는 일반적인 방법으로 얻어오는 것 말고 circuitbreaker를 위한 설정파일을 추가해보자.


## `Resilience4jConfig.java`
Circuit Breaker를 커스터마이징 하기 위해서 해당 클래스를 생성한다.
```java
@Configuration
public class Resilience4jConfig {

    @Bean
    public Customizer<Resilience4JCircuitBreakerFactory> globalCustomConfiguration() {

        CircuitBreakerConfig circuitBreakerConfig = CircuitBreakerConfig.custom()
            .failureRateThreshold(4)
            .waitDurationInOpenState(Duration.ofMillis(1000))
            .slidingWindowType(CircuitBreakerConfig.SlidingWindowType.COUNT_BASED)
            .slidingWindowSize(2)
            .build();

        TimeLimiterConfig timeLimiterConfig = TimeLimiterConfig.custom()
            .timeoutDuration(Duration.ofSeconds(4))
            .build();

        return factory -> factory.configureDefault(id -> new Resilience4JConfigBuilder(id)
            .timeLimiterConfig(timeLimiterConfig)
            .circuitBreakerConfig(circuitBreakerConfig)
            .build()
        );
    }
}
```

<img width="1109" alt="스크린샷 2022-10-13 오전 10 42 15" src="https://user-images.githubusercontent.com/59405576/195479156-e0376fb3-12fb-4e30-94fa-43c22070bdb7.png"><br><br>
<img width="1039" alt="스크린샷 2022-10-13 오전 10 42 40" src="https://user-images.githubusercontent.com/59405576/195479218-f45f4c36-9012-4709-817b-071661f52a9c.png">





***
<br>


    💛 개인 공부 기록용 블로그입니다. 👻

[맨 위로 이동하기](#){: .btn .btn--primary }{: .align-right}